Add HNSW ACE build method

julianmi opened the pull request:
This PR adds a direct `hnsw::build` API that uses the ACE (Augmented Core Extraction) algorithm to build HNSW indexes on the GPU. ACE enables building HNSW indexes for datasets too large to fit in GPU memory by partitioning the data and building sub-indexes.

CC @tfeher 

**C++ API**
- Added `hnsw::build()` function with ACE parameters for direct HNSW index construction. This serializes an HNSW index to disk if `use_disk` is true.
- Added `hnsw::graph_build_params::ace_params` struct with configurable options:
  - `npartitions` - number of partitions for parallel build
  - `ef_construction` - index quality parameter
  - `build_dir` - directory for disk-based build artifacts
  - `use_disk` - force disk-based storage mode
- Implemented proper serialization/deserialization for disk-backed HNSW indexes
- Added C++ tests in `ann_hnsw_ace.cuh`

**C API**
- Added `cuvsHnswBuild` function with ACE parameters
- Added C tests in `ann_hnsw_ace.cu`

**Python**
- Added `hnsw.AceParams` class for configuring ACE builds
- Added Python tests in `test_hnsw_ace.py`

**Java**
- Added `HnswAceParams` class
- Added Java tests in `HnswAceBuildAndSearchIT.java`

**Documentation**
- Added `cuvs_hnsw` section to the parameter tuning guide with ACE parameters

**Example**
- Added `hnsw_ace_example.cu` demonstrating the build → deserialize → search workflow

julianmi committed:
Add HNSW ACE build method

- Added `cuvsHnswAceParams` structure for ACE configuration.
- Implemented `cuvsHnswBuild` function to facilitate index construction using ACE.
- Updated HNSW index parameters to include ACE settings.
- Created new tests for HNSW index building and searching using ACE.
- Updated documentation to reflect the new ACE parameters and usage.

julianmi committed:
Fix formatting

tfeher reviewed (changes requested):
Thanks Julian for the PR! I think it is a nice improvement, but I would still recommend to simplify further along the following lines
- Treat CAGRA parameters as implementation details. We want to keep the usage and the documentation simple for users familiar with HNSW index building.
- Still provide a way to adjust these parameters.
- Clarify that the results are not an exact equivalent to a HNSW graph, but in practice the graph is a good replacement that works well in HNSW search.

julianmi committed:
Address review feedback

julianmi commented on cpp/include/cuvs/neighbors/hnsw.hpp:
```diff
@@ -51,6 +56,30 @@ struct index_params : cuvs::neighbors::index_params {
+   *
+   * Set ace_params to configure ACE (Augmented Core Extraction) parameters for building
+   * a GPU-accelerated HNSW index. ACE enables building indexes for datasets too large
+   * to fit in GPU memory.
```
Done.

julianmi commented on examples/cpp/src/hnsw_ace_example.cu:
```diff
@@ -51,90 +54,65 @@ void cagra_build_search_ace(raft::device_resources const& dev_resources,
+
+  // For disk-based indexes, the build function serializes the index to disk
+  // We need to deserialize it before searching
+  std::string hnsw_index_path = "/tmp/hnsw_ace_build/hnsw_index.bin";
```
Yes, with `hnsw_index->file_path()`. I have changed this.

julianmi commented on examples/cpp/src/hnsw_ace_example.cu:
```diff
@@ -11,34 +11,37 @@
-  // CAGRA index parameters
-  cagra::index_params index_params;
-  index_params.intermediate_graph_degree = 128;
-  index_params.graph_degree              = 64;
```
Done.

julianmi commented on examples/cpp/src/hnsw_ace_example.cu:
I have restored `cagra_hnsw_ace_example.cu`. I kept `npartitions` for now. I will remove it in the next PR adding the heuristic.

julianmi commented on cpp/include/cuvs/neighbors/hnsw.hpp:
```diff
@@ -158,6 +187,107 @@ struct extend_params {
+  raft::host_matrix_view<const float, int64_t, raft::row_major> dataset);
+
+/**
+ * @brief Build an HNSW index using the ACE algorithm (half precision)
```
Good point. I have copied the full docstring here and below.

julianmi committed:
Improve C++ test based on review feedback

tfeher commented on cpp/include/cuvs/neighbors/hnsw.hpp:
```diff
@@ -51,6 +56,31 @@ struct index_params : cuvs::neighbors::index_params {
+   * params.graph_build_params = hnsw::graph_build_params::ace_params();
+   * auto& ace = std::get<hnsw::graph_build_params::ace_params>(params.graph_build_params);
+   * ace.npartitions = 4;
+   * ace.ef_construction = 120;
```
We have `hnsw::index_params::ef_construction` and `ace_params::ef_construction`. We need to clarify which one to use. I recommend to use `hnsw::index_params::ef_construction`.

tfeher commented on python/cuvs/cuvs/neighbors/hnsw/hnsw.pyx:
```diff
@@ -50,20 +116,34 @@ cdef class IndexParams:
         NOTE: When hierarchy is `gpu`, while the majority of the work is done
         on the GPU, initialization of the HNSW index itself and some other
         work is parallelized with the help of CPU threads.
+    m : int, default = 32 (optional)
```
```suggestion
    M : int, default = 32 (optional)
```

tfeher reviewed (changes requested):
Thanks @julianmi for the updates! There are only two issues remaining:
1. Redundancy with `ef_construction` parameter (see comments below)
2. Build - deserialize - search workflow

Considering the second point, currently the resulting index can be either in memory or stored in a file. (Although #1604 removes the in-memory ACE path, we might still decide later to enable other in-memory build methods.)
```c++
auto hnsw_index = hnsw::build(res, hnsw_params, dataset);
// index is saved to disk at hnsw_index->file_path(), the hnsw_index structure just a wrapper around the file name and index params
hnsw::search(res, search_params, *hnsw_index, queries, indices, distances);
```
This would fail with the following error message if the index does not fit memory. 
```
Searching HNSW index
terminate called after throwing an instance of 'raft::logic_error'
  what():  RAFT failure at file=/home/scratch.tfeher_gpu_2/cuvs_1597/cpp/src/neighbors/detail/hnsw.hpp line=1140: Cannot search an HNSW index that is stored on disk. The index must be deserialized into memory first using hnsw::deserialize().
Obtained 7 stack frames
```

Instead, when the index is on disk, then we need to use the following step
```c++
auto hnsw_index = hnsw::build(res, hnsw_params, dataset);
hnsw::deserialize(res, hnsw_params, hnsw_index->file_path(), dataset.extent(1), hnsw_params.metric, &hnsw_index_deserialized);
hnsw::search(res, search_params, *hnsw_index_deserialized, queries, indices, distances);
```

Do we have to dictate such usage patterns? Could we instead let `hnsw::search` load the index from file if needed?

julianmi committed:
Merge branch 'main' into hnsw-ace

julianmi committed:
Address review comments

- Renamed parameter `m` to `M` in HNSW structures and related functions for consistency.
- Removed `ef_construction` from `cuvsHnswAceParams` and related classes, as it is no longer needed.
- Load the HNSW index from file before search if needed.

julianmi commented on cpp/include/cuvs/neighbors/hnsw.hpp:
```diff
@@ -51,6 +56,31 @@ struct index_params : cuvs::neighbors::index_params {
+
+  /** HNSW M parameter: number of bi-directional links per node (used when building with ACE).
+   */
+  size_t m = 32;
```
Capitalized `M` throughout.

julianmi commented on cpp/src/neighbors/detail/hnsw.hpp:
```diff
@@ -1196,12 +1212,69 @@ void deserialize(raft::resources const& res,
+  // Configure ACE parameters for CAGRA
+  cuvs::neighbors::cagra::graph_build_params::ace_params cagra_ace_params;
+  cagra_ace_params.npartitions     = ace_params.npartitions;
+  cagra_ace_params.ef_construction = ace_params.ef_construction;
```
I have removed `ace_params.ef_construction` altogether.

tfeher reviewed (approved):
Thanks @julianmi for the updates, the PR looks good to me!

julianmi committed:
Merge branch 'main' into hnsw-ace

julianmi committed:
Merge branch 'main' into hnsw-ace

julianmi committed:
Merge branch 'main' into hnsw-ace

julianmi committed:
Merge branch 'main' into hnsw-ace

julianmi committed:
Merge branch 'main' into hnsw-ace

julianmi committed:
Align HNSW ACE Python test with CAGRA ACE test

julianmi committed:
Merge branch 'main' into hnsw-ace

divyegala reviewed (approved):
Python approval only

julianmi committed:
Merge branch 'main' into hnsw-ace

julianmi committed:
Merge branch 'main' into hnsw-ace

julianmi committed:
Merge branch 'main' into hnsw-ace

julianmi committed:
Update copyright

julianmi committed:
Merge branch 'main' into hnsw-ace

julianmi committed:
Merge branch 'main' into hnsw-ace

tfeher committed:
Merge branch 'main' into hnsw-ace

julianmi committed:
Merge branch 'main' into hnsw-ace

mythrocks reviewed (approved):
Approving the Java bit.

tfeher commented:
/merge
